// griddle/gulp-project.js
var fs = require('fs');
var q = require('q');
var util = require('util');
var path = require('path');
var cp = require('child_process');
var events = require('events');

var SerialQueue = require('./serial-queue');

function GulpProject(src) {
  this.src = path.resolve(src);
  if(!this.src || !fs.existsSync(this.src) || !fs.statSync(this.src).isDirectory()) {
    throw util.format('Invalid directory: "%s"', src);
  }
  this.queue = new SerialQueue();
}

util.inherits(GulpProject, events.EventEmitter);

GulpProject.prototype.status = function() {
  if(this.queue.errors.length > 0) {
    return GulpProject.Status.LAST_BUILD_FAILED;
  } else if(this.queue.status === SerialQueue.Status.EXECUTING) {
    return GulpProject.Status.BUILDING;
  }
  return GulpProject.Status.IDLE;
};

GulpProject.prototype.build = function() {
  if(this.status() === GulpProject.Status.LAST_BUILD_FAILED) {
    this.queue.reset();
  }
  this.queue.push(function() {
    var built = q.defer();
    cp.exec('gulp', { cwd: this.src }, function(err, stdout, stderr) {
      if(err || stderr) {
        built.reject([ stdout, stderr ].join('\n'));
      } else {
        built.resolve();
      }
    });
    return built.promise;
  }.bind(this));

  return this;
};

GulpProject.prototype.buildCompleted = function() {
  var p = q.defer();
  if(this.queue.status === SerialQueue.Status.IDLE) {
    p.resolve();
  } else {
    this.queue.once(SerialQueue.Event.DONE, function() {
      if(this.status() === GulpProject.Status.LAST_BUILD_FAILED) {
        p.reject();
      } else {
        p.resolve();
      }
    }.bind(this));
  }
  return p.promise;
};

GulpProject.prototype.log = function() {
  return this.queue.log.filter(function(entry) {
    // Filter empty strings.
    return entry && entry.length > 0;
  }).join('\n');
};

GulpProject.Event = {
  BUILD_SUCCESSFUL: 'build-successful',
  BUILD_FAILED: 'build-failed'
};
Object.freeze(GulpProject.Event);

GulpProject.Status = {
  IDLE: 'idle',
  BUILDING: 'building',
  LAST_BUILD_SUCCESSFUL: 'last-build-successful',
  LAST_BUILD_FAILED: 'last-build-failed'
};
Object.freeze(GulpProject.Status);

module.exports = GulpProject;
